controllers/users_config_map_controller.go

```diff
@@ -0,0 +1,281 @@
+	log := ctrllog.FromContext(ctx).WithName("ConfigMap")
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	// fetch the config map instance
+	configMap := &corev1.ConfigMap{}
+	if err := r.Client.Get(ctx, req.NamespacedName, configMap); err != nil {
+		if errors.IsNotFound(err) {
+			// The object is deleted, nothing to do
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "failed to get config map", l.Action, l.ActionView)
+		return ctrl.Result{}, err
+	}
+
+	if !controllerutil.ContainsFinalizer(configMap, ConfigMapFinalizer) {
+		controllerutil.AddFinalizer(configMap, ConfigMapFinalizer)
+
+		if err := r.Client.Update(ctx, configMap); err != nil {
+			log.Error(err, "failed to add finalizer to config map", "ConfigMapName", additionalUsersConfigMapName, "Finalizer", ConfigMapFinalizer, l.Action, l.ActionUpdate)
+			return ctrl.Result{}, err
+		}
+		log.Info("finalizer added to configmap")
+		// leave all other steps to reconcile from update action
+		return ctrl.Result{}, nil
+	}
+
+	teamName := getQuayTeamName(configMap.Namespace)
+	removeTeam := false
+
+	if !configMap.DeletionTimestamp.IsZero() {
+		removeTeam = true
+		log.Info("Config map with additional users is being removed, will delete team", "TeamName", teamName)
+	}
+
+	var additionalUsers []string
+	if !removeTeam {
+		// get additional users from config map
+		additionalUsersStr, usersExist := configMap.Data[additionalUsersConfigMapKey]
+		if !usersExist {
+			log.Info("Config map with additional users doesn't have the key", "ConfigMapName", additionalUsersConfigMapName, "ConfigMapKey", additionalUsersConfigMapKey, l.Action, l.ActionView)
+			removeTeam = true
+		} else {
+			additionalUsers = strings.Fields(strings.TrimSpace(additionalUsersStr))
+			log.Info("Additional users configured in config map", "AdditionalUsers", additionalUsers)
+		}
+	}
+
+	// reread quay token
+	r.QuayClient = r.BuildQuayClient(log)
+
+	// remove team if config map is being removed, or doesn't contain key additionalUsersConfigMapKey
+	if removeTeam {
+		log.Info("Will remove team", "TeamName", teamName)
+		if err := r.QuayClient.DeleteTeam(r.QuayOrganization, teamName); err != nil {
+			log.Error(err, "failed to remove team", "TeamName", teamName, l.Action, l.ActionDelete)
+			return ctrl.Result{}, err
+		}
+
+		// remove finalizer after team is removed
+		if !configMap.DeletionTimestamp.IsZero() {
+			if controllerutil.ContainsFinalizer(configMap, ConfigMapFinalizer) {
+				controllerutil.RemoveFinalizer(configMap, ConfigMapFinalizer)
+				if err := r.Client.Update(ctx, configMap); err != nil {
+					log.Error(err, "failed to remove config map finalizer", "ConfigMapName", additionalUsersConfigMapName, "Finalizer", ConfigMapFinalizer, l.Action, l.ActionUpdate)
+					return ctrl.Result{}, err
+				}
+				log.Info("finalizer removed from config map", l.Action, l.ActionDelete)
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	// fetch image repositories in the namespace
```

Comment:
Since we use only image repositories names further, does it make sense to have a separate function that just returns the list?

Reply:
created new function